prepare.pyx

```diff
@@ -240,16 +250,56 @@ cpdef inline prepare_time_micros(object data):
 
 
 
-def prepare_nano_time(data):
-	if isinstance(data, qutils.time.nanotime.NanoTime):
-		return data.nanoseconds
-	else:
-		return data
+cdef prepare_nano_time(data):
+	return data.nanoseconds
 
 
 
-def prepare_nano_time_delta(data):
-	if isinstance(data, qutils.time.NanoTimeDelta):
-		return data.nanoseconds
-	else:
-		return data
\ No newline at end of file
+cdef prepare_nano_time_delta(data):
+	return data.nanoseconds
+
+
+
+cdef inline read_timestamp_millis(data):
+	return parse_timestamp(data, float(1000))
+
+cdef inline read_timestamp_micros(data):
+	return parse_timestamp(data, float(1000000))
+
+cdef inline read_date(data):
+	return datetime.date.fromordinal(data + datetime.date(1970, 1, 1).toordinal())
+
+cdef inline read_uuid(data):
+	return uuid.UUID(data)
+
+cdef inline read_decimal(data, schema):
+	scale = schema.get('scale', 0)
+	precision = schema['precision']
+	unscaled_datum = int.from_bytes(data, byteorder = 'big', signed = True)
+	decimal_context = decimal.Context()
+	decimal_context.prec = precision
+	return decimal_context.create_decimal(unscaled_datum). \
+		scaleb(-scale, decimal_context)
+
+cdef inline read_time_millis(data):
+	h = int(data / 60 * 60 * 1000)
+	m = int(data / 60 * 1000) % 60
+	s = int(data / 1000) % 60
+	mls = int(data % 1000) * 1000
+	return datetime.time(h, m, s, mls)
+
+cdef inline read_time_micros(data):
+	h = int(data / (60 * 60 * 1000000))
+	m = int(data / (60 * 1000000)) % 60
+	s = int(data / 1000000) % 60
+	mcs = data % 1000000
+	return datetime.time(h, m, s, mcs)
+
+cdef inline parse_timestamp(data, resolution):
+	return datetime.datetime(1970, 1, 1, tzinfo = utc) + datetime.timedelta(seconds = data / resolution)
+
+cdef read_nano_time(data):
+	return qutils.time.nanotime.NanoTime(data)
+
+cdef read_nano_time_delta(data):
+	return qutils.time.NanoTimeDelta(data)
```

Comment:
Will this compile if we don't have `qutils` installed? Have you tried it?

Reply:
Yep, I tried it both locally and in the CI here at Github. It seems weird to me but its really really nice that it behaves like this.